tutorials/learn_digital_audio/04_Filter.cpp

```diff
@@ -0,0 +1,139 @@
+// Joel A. Jaffe 2024-07-18
+
+/* Filter
+This app demonstrates how to enable user input, and also how to generate 
+an impulse response for testing audio objects like filters.
+
+We've created a bipolar oscillator object that can produce various elemental waveforms,
+however have not sonified them at audio rates (frequencies > ~20Hz). 
+This is because the square, sawtooth, and triangle waveforms 
+contain frequencies above half the sample rate, and therefore will generate 
+a type of digital distortion known as 'aliasing'.
+
+To fix this, we can use a digital filter to remove the high frequency content.
```

Comment:
Note that a digital filter will not fix aliasing once it's occured. The only way to combat aliasing is prevention. There is a special class of oscillators that start with the waveform integral and then numerically differentiate it which introduces a zero at Nyquist. The original paper called these DPW (differentiated parabolic waves), but it is a much more general technique. Another technique is to create a series of bandlimited wavetables and then cross-fade between them according to oscillator frequency. This is the same technique used to prevent aliasing in texture lookups in graphics called mipmapping.

Reply:
Sounds like I have some reading to do! If you could provide a name/link to the DPW paper I would greatly appreciate it. I will revisit the mipmapping section of Graham's book today. Thank you!

Reply:
Here is a link to one of the papers (you can then find others in its orbit): https://mac.kaist.ac.kr/pubs/ValimakiNamSmithAbel-taslp2010.pdf

In Gamma, I've called it a DWO (differenced wave oscillator): https://github.com/LancePutnam/Gamma/blob/b508d7ab50f5316d26b4e521c998a9a8b34a7917/Gamma/Oscillator.h#L753

Reply:
Small correction on my statement: The first-order differencing used by DPW introduces a zero at DC not Nyquist! So, in effect it is a high-pass filter. I believe some synth manufacturers (Roland?) used a similar technique with their digital oscillators by sending them through a high-pass which tracked the fundamental. That way you live with some aliasing, but at least suppress the most offensive partials at low frequencies.

Reply:
Hi Lance!

After surveying the list of anti-aliasing techniques presented by Graham, I've concluded that DPW is the best option for this project- per your initial suggestion. I prototyped using the following codeblock and observed (audibly and visually) the best balance of anti-aliasing and hf rolloff with the filter's $\alpha$ value around 0.75

```
virtual T processSample() override {
    T nextPhase = Phasor<T>::processSample();
    T diff = nextPhase - prevPhase;
    prevPhase = nextPhase;
    T output = filter.lpf(diff) + prevOutput;
    prevOutput = output;
    return 2 * output - 1; 
  }
```

Does the above logic and implementation sit soundly with you? Looking forward to integrating the changes and submitting a new PR. 

Many thanks, 
Joel

Reply:
Good choice. Here is a simple AA saw example inspired by DPW:
```
class SawAA {
public:

	void freq(float v){
		mFreq = v<-0.5f ? -0.5f : v>0.5f ? 0.5f : v;
		mAmp = 0.25f/std::abs(mFreq);
	}

	float operator()(){
		// Get previous and current phase
		auto prevPhase = mPhase;
		mPhase += mFreq;
		mPhase = mPhase - std::floor(mPhase);

		// Compute integral waves (parabolas)
		auto pow2 = [](float x){ return x*x; };
		auto i1 = pow2(2.f*prevPhase - 1.f);
		auto i2 = pow2(2.f*mPhase    - 1.f);

		// Return difference (saw)
		return (i2 - i1) * mAmp;
	}

private:
	float mPhase = 0.f;
	float mFreq = 0.f;
	float mAmp = 1.f;
};
```
Why inspired? Because there is no filter to speak of. This is a purely analytic solution. In my experience, the differencing filter in the original DPW algorithm can misbehave under audio rate FM since it affects the amplitude of the signal. Here, we only work with phase. Most basic waveforms are cheap enough to compute on the fly, so performance should not be much worse than using a filter (if your CPU is smart it should pipeline those instructions).

Reply:
Very cool!  I couldn't help it, I ported it to gen~ to compare with some other methods.  Here's the spectrum plot of this DPW-inspired patch compared with the box-integral patch from the GO book -- here at 2862Hz at a SR of 48kHz. They look almost identical -- with the aliasing folding around and going down to zero at SR/0Hz. 

<img width="658" alt="Screenshot 2024-08-07 at 10 33 14 AM" src="https://github.com/user-attachments/assets/8b31d622-e38c-4295-935b-14f87d4b7f45">

Here are the patches. DPW-inspired:

<img width="330" alt="Screenshot 2024-08-07 at 10 35 24 AM" src="https://github.com/user-attachments/assets/33eabd3a-dc02-40ac-8008-d69645eebabe">

Box-integral inspired:

<img width="509" alt="Screenshot 2024-08-07 at 10 35 57 AM" src="https://github.com/user-attachments/assets/d14f9349-bfb5-4140-96dc-179c7dc65cdf">

The idea of the box integral is that you are running two phasors projected half a sample before and after the true phase, and cross fade between them centred at the subsample transition point of the saw step.

This box integral method is equivalent to a very poor BLEP of a linear ramp over 1 sample. The polyblep method does a better job:

<img width="642" alt="Screenshot 2024-08-07 at 10 43 24 AM" src="https://github.com/user-attachments/assets/d2e1b6c7-659b-4a95-af7e-760d1900c9cf">

The algorithm is also pretty simple, would be easy to port this to C++:

<img width="507" alt="Screenshot 2024-08-07 at 10 46 40 AM" src="https://github.com/user-attachments/assets/6f4ed5af-2a8d-455e-b4b1-64f74dbb645f">

Reply:
Rough cpp translation of the polyblep saw (untested):

```c++


float blepsaw() {
  // raw ramp:
  auto oldphase = mPhase;
  mPhase += mFreq;
  mPhase -= floor(mPhase); 
  
  // raw bipolar saw:
  auto saw = mPhase * 2.f - 1.f;
  
  // compute the blep correction:
  auto b = mPhase - 0.5f;
  auto blep = 0.5f - std::abs(b); // upside-down triangle wave from 0.0 to 0.5
  blep /= std::abs(mFreq); // subsample time since ramp step
  blep = std::min(1.f, std::max(0.f, blep)); // we only care if the step happens this sample
  blep = (blep - 1.f) * (1.f - blep); // the blep polynomial correction
  blep = (b >= 0.f) ? blep : -blep; // flip the correction for 2nd half of transition
  
  // add to raw saw:
  return raw + blep;
}


```

Reply:
Thank you both! It was especially helpful to see the two methods both as C++ and Gen~. 

I've modified the tutorial series to include `04_Aliasing`, which implements an `AntiAliasedOsc` that inherits from `SwissArmyOsc`, which has had it's waveshaping logic encapsulated in a function separate from `processSample()`. 

At first I dropped the waveshaping logic into Lance's DPW-inspired design, where his saw logic was: 

```
auto i1 = pow2(this->waveshape(prevPhase));
auto i2 = pow2(this->waveshape(nextPhase);
return (i2 - i1) * (0.25/phaseIncrement);
```

The resultant saw (as expected) and sine waveforms looked right, but 0 was returned for the square wave, and the triangle wave returned an amplified saw.

I tried the blep implementation as an alternative, substituting my waveshaping logic in the definition of `raw`. All waveforms had the appropriate shape, but exhibited (except for saw) discontinuity artifacts at the wrapping point. 

Can either of these methods be made to "play nice" with the substitution of various waveforms?

Many thanks,
Joel

Reply:
Did the blep saw come out anti-aliased? The C++ code I posted is untested, just an attempt to translate the gen~ patch. I think the C++ is probably right but I might have made an error somewhere. You should verify that it works first. 

I couldn't find `04_Aliasing` in your fork, so I'm not sure how you are deriving the other shapes.   

For example, you can't get an antialiased pulse by simply sending an antialiased saw through a > 0 comparator. You'd have to antialias the actual comparator instead.  Similarly you can't just send an antialiased saw through an arbitrary waveshaper function and assume an antialiased output; it depends very much what the waveshaper function looks like -- if it is low order polynomial with no discontinuities you can probably get away with it, but if there are any discontinuities then definitely not. Those discontinuities also need to be antialiased. 

But you can derive an antialiased pulse wave by subtracting two antialiased saw waves that are phase shifted; the phase shift between them defines the pulse width. You also need to add (or subtract?) the pulse width to get them centered around DC=0.   

Or you could just put blep corrections where the pulse steps should be, which is essentially the same as antialiasing a comparator.

Reply:
As far as I can tell, the blep and DPW-inspired techniques both work as intended for the saw. 
I sent the signal from my app into Max to confirm.

DPW-inspired result(Hz = 2862):
<img width="437" alt="putnam" src="https://github.com/user-attachments/assets/6973ce2e-b4d6-4d00-bcae-072d15617171">

Blep result(Hz = 2862):
<img width="436" alt="blep" src="https://github.com/user-attachments/assets/4e20c0f3-b35a-4ae7-9ebd-7a0a0cdc0c19">

I just pushed `04_Aliasing`, which shows where the waveshaping logic is applied and how it interacts with the two anti-aliasing techniques. The waveform, freq, etc can be adjusted by playing with the constructor for `AntiAliasedOsc`, and the anti-aliasing technique selected by commenting/uncommenting of the `return` statement for `processSample()`

Please excuse the hacky-ness of it, I just wanted to draft a quick tool for comparison of the techniques. 
Thanks again for the generous help.

Reply:
As Graham says, you cannot always resort to "business as usual" on the AAed waves and expect the results to be AAed. The key MO with all of these techniques is to disallow jump discontinuites and corners. Graham's suggestion of subtracting two AAed saws for a pulse is a good one. You can use any linear combinations of AAed waves and remain safe. Now, for getting more waves out of DPW, you just need to work out what the integral wave is of the waveform you want through some calculus or whatever. Here is the DPW C++ extended to include other oscillators and an arbitrary function:
```
/// Anti-aliased oscillator using difference of integral waves

/// Some of the integral waves used can be found here:
/// https://www.desmos.com/calculator/gez7vmyrol
class OscAA {
public:

	void freq(float v){
		mFreq = v<-0.5f ? -0.5f : v>0.5f ? 0.5f : v;
		// Prevent freq from getting too close to zero
		float eps = 1e-8;
		if(mFreq>-eps && mFreq<eps) mFreq = mFreq>0.f ? eps : -eps;
		mAmp = 0.25f/std::abs(mFreq);
	}

	/// Returns anti-aliased derivative of passed in wave function

	/// \param[in] iwave	Integral of desired wave---input domain is [0,1) and
	///						output range should be in [-1/1, 1/2].
	template <class IntegralWave>
	float operator()(const IntegralWave& iwave){
		// Get previous and current phase
		auto prevPhase = mPhase;
		mPhase += mFreq;
		mPhase = mPhase - std::floor(mPhase);

		// Compute integral waves
		auto i1 = iwave(prevPhase);
		auto i2 = iwave(mPhase   );

		// Return difference
		return (i2 - i1) * mAmp;
	}

	float saw(){
		return (*this)([](float p){
			// parabolic wave in [0,1]
			p = 2.f*p - 1.f;
			return p*p;
		});
	}

	float square(){
		return (*this)([](float p){
			// triangle wave in [0,1]
			return std::abs(2.f*p - 1.f);
		});
	}

	float para(){
		return (*this)([](float p){
			// Clausen wave in [0,1]
			static constexpr float c = std::sqrt(3.)*0.75;
			p = 2.f*p-1.f; // to saw in [-1,1]
			return c*(p*p*p-p); // output in [-1/2, 1/2]
		});
	}

	float triangle(){
		return (*this)([](float p){
			// "parabolic sine" wave in [0,1]
			p = 2.f*p-1.f; // to saw in [-1,1]
			return 2.f*p*(std::abs(p) - 1.f); // output in [-1/2, 1/2]
		});
	}

private:
	float mPhase = 0.f;
	float mFreq = 0.f;
	float mAmp = 1.f;
};
```
I added a guard to prevent small frequencies as they will cause the normalization amplitude to blow up. That said, low frequencies can present a numerical issue with DPW as you end up dividing a very small number by a very large one. This might be one advantage BLEP has over DPW.

Reply:
Thank you Lance! This really made it click for me.
I moved over to gen~ for prototyping, and am very happy with the results (below)!
As I integrate this band-limiting technique into my tutorial series, I'd love to hear your thoughts on how to concisely introduce and explain it. 

<img width="712" alt="AAgen" src="https://github.com/user-attachments/assets/e64d3212-7ffd-4886-9189-ab00cbe1d613">

<img width="485" alt="AA Saw" src="https://github.com/user-attachments/assets/c79b3646-4ba4-422b-ad9f-020230f84125">

<img width="489" alt="AASqr" src="https://github.com/user-attachments/assets/e16696e3-2b09-4696-91f3-3cf10ff8bd67">

<img width="495" alt="AATri" src="https://github.com/user-attachments/assets/ea3a40cb-3118-4a79-ac47-2b490ee26df7">

Reply:
No problem. Note, I made some small corrections to the code above. The output integral waveforms _do not_ need to be in [-1/2, 1/2], but rather simply be the analytic integral of the desired waveform. The parabolic wave had the wrong amplitude, but now it is fixed to have a peak amplitude of 1.

To introduce, I would say that DPW falls into a class of band-limiting techniques which include BLEP and others mentioned here. As stated in the paper above, the intuition behind DPW is to start with a waveform with acceptable aliasing and then apply a high-pass differencing filter to brighten it and reject any aliased partials below the fundamental. There is some consensus that Roland used such a high-pass approach on their JP8000 synth which was released in 1997: https://edouard.digital/jp-booo/. My improvement to DPW eliminates a problem with amplitude fluctuations and popping in the original DPW algorithm when the frequency changes. It replaces the differencing (one-zero) filter with an instantaneous numerical difference between two phase adjusted waves. By eliminating the filter, there is no longer a dependency on the previous sample (which can throw off the normalizer) and one no longer needs to be concerned about how to initialize the filter.

Reply:
Your elegant simplification of the algorithm made it much easier for me to wrap my head around, and hopefully will in turn make the tutorial digestible for others. Perhaps in time the "Putnamian Oscillator" will become a staple of digital synthesis! 

With gratitude,
Joel